File: client/src/main/java/io/github/hulang1024/chinesechess/scene/chessplay/rule/chess/ChessS.java
package io.github.hulang1024.chinesechess.scene.chessplay.rule.chess;

import io.github.hulang1024.chinesechess.scene.chessplay.rule.ChessPosition;
import io.github.hulang1024.chinesechess.scene.chessplay.rule.HostEnum;
import io.github.hulang1024.chinesechess.scene.chessplay.rule.RoundGame;

/**
 * 兵
 * @author Hu Lang
 */
public class ChessS extends AbstractChess {
    public ChessS(ChessPosition pos, HostEnum host) {
        super(pos, host);
    }

    @Override
    public boolean canGoTo(ChessPosition destPos, RoundGame game) {
        int colOffset = Math.abs(destPos.col - pos.col);

        // 是否向前单步
        boolean isForward = colOffset == 0
            && (game.isHostAtChessboardTop(host)
                ? pos.row + 1 == destPos.row
                : pos.row - 1 == destPos.row);

        // 判断是否过了河
        if (MoveRules.isInBoundary(game, host, destPos)) {
            // 过河之前只可以向前单步
            return isForward;
        } else {
            // 过河之后既可以向前单步，也可以左或右移单步
            return isForward || colOffset == 1;
        }
    }
}

File: client/src/rule/chess/ChessS.ts
import ChessPos from "../ChessPos";
import Game from "../Game";
import AbstractChess from "./AbstractChess";
import { isInBoundary } from "./move_rules";

/**
 * 兵
 */
export default class ChessS extends AbstractChess {
  canGoTo(destPos: ChessPos, game: Game): boolean {
    const rowOffset = destPos.row - this.pos.row;
    const colOffset = destPos.col - this.pos.col;

    // 是否向前单步
    const isForward = Math.abs(colOffset) == 0
      && (game.isHostAtChessboardTop(this.host) ? rowOffset == +1 : rowOffset == -1);

    // 判断是否过了河
    if (isInBoundary(game, this.host, destPos)) {
      // 过河之前只可以向前单步
      return isForward;
    }
    // 过河之后既可以向前单步，也可以左或右移单步
    return isForward || (Math.abs(rowOffset) == 0 && Math.abs(colOffset) == 1);
  }
}
